src/main/java/com/conveyal/gtfs/stats/PatternStats.java

```diff
@@ -1,7 +1,176 @@
 package com.conveyal.gtfs.stats;
 
+import com.conveyal.gtfs.GTFSFeed;
+import com.conveyal.gtfs.model.Pattern;
+import com.conveyal.gtfs.model.StopTime;
+import com.conveyal.gtfs.model.Trip;
+import gnu.trove.list.TDoubleList;
+import gnu.trove.list.TIntList;
+import gnu.trove.list.array.TDoubleArrayList;
+import gnu.trove.list.array.TIntArrayList;
+import org.mapdb.Fun;
+
+import java.time.LocalDate;
+import java.time.LocalTime;
+import java.util.Collection;
+import java.util.List;
+import java.util.Spliterator;
+import java.util.stream.Collectors;
+import java.util.stream.StreamSupport;
+
 /**
  * Created by landon on 9/2/16.
  */
 public class PatternStats {
+
+    private GTFSFeed feed = null;
+    private FeedStats stats = null;
+
+    public PatternStats (GTFSFeed f, FeedStats fs) {
+        feed = f;
+        stats = fs;
+    }
+
+    public double getPatternSpeed (String pattern_id, LocalDate date, LocalTime from, LocalTime to) {
+
+        List<Trip> trips = getTripsForDate(pattern_id, date);
+
+        return getAverageSpeedForTrips(trips, from, to);
+    }
+
+    public double getAverageSpeedForTrips (Collection<Trip> trips, LocalTime from, LocalTime to) {
+        TDoubleList speeds = new TDoubleArrayList();
+
+        for (Trip trip : trips) {
+            StopTime firstStopTime = feed.stop_times.ceilingEntry(Fun.t2(trip.trip_id, null)).getValue();
+            LocalTime tripBeginTime = LocalTime.ofSecondOfDay(firstStopTime.departure_time % 86399); // convert 24hr+ seconds to 0 - 86399
+
+            // skip trip if begin time is before or after specified time period
+            if (tripBeginTime.isAfter(to) || tripBeginTime.isBefore(from)) {
+                continue;
+            }
+            // TODO: swap straight lines for actual geometry?
+            double speed = feed.getTripSpeed(trip.trip_id, true);
+
+            if (!Double.isNaN(speed)) {
+                speeds.add(speed);
+            }
+        }
+
+        if (speeds.isEmpty()) return Double.NaN;
+
+        return speeds.sum() / speeds.size();
+    }
+
+    public LocalTime getStartTimeForTrips (Collection<Trip> trips) {
+        int earliestDeparture = Integer.MAX_VALUE;
+
+        for (Trip trip : trips) {
+            StopTime st;
+            try {
+                // use interpolated times in case our common stop is not a time point
+                st = StreamSupport.stream(feed.getInterpolatedStopTimesForTrip(trip.trip_id).spliterator(), false)
+                        .findFirst()
+                        .orElse(null);
+            } catch (GTFSFeed.FirstAndLastStopsDoNotHaveTimes e) {
+                return null;
+            }
+
+            // shouldn't actually encounter any departures after midnight, but skip any departures that do
+            if (st.departure_time > 86399) continue;
```

Comment:
Believe this is actually legal but it's a trip that runs the next morning, so I'd say ignoring it is fine, but might clarify the comment.